lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -849,14 +883,14 @@ LogicalResult EQPrimOp::canonicalize(EQPrimOp op, PatternRewriter &rewriter) {
 OpFoldResult NEQPrimOp::fold(FoldAdaptor adaptor) {
   // neq(x, x) -> 0
   if (getLhs() == getRhs())
-    return getIntAttr(getType(), APInt(1, 0));
+    return foldResult(*this, getIntAttr(getType(), APInt(1, 0)));
 
   if (auto rhsCst = getConstant(adaptor.getRhs())) {
     /// neq(x, 0) -> x when x is 1 bit.
     /// TODO: Support SInt<1> on the LHS etc.
-    if (rhsCst->isZero() && getLhs().getType() == getType() &&
-        getRhs().getType() == getType())
-      return getLhs();
+    if (rhsCst->isZero() && mixedConstTypes(getLhs().getType(), getType()) &&
+        mixedConstTypes(getRhs().getType(), getType()))
+      return foldResult(*this, getLhs());
```

Comment:
Could we table-gen wrap our fold methods with logic that calls this for us?

Mostly, this means we have to be "good" about always doing this when writing any fold logic, which seems easy to overlook.  Just a thought.

Reply:
I was hoping to find a way to automatically do this but couldn't find anything. If you have any ideas that'd be great.

Reply:
Thanks for the ideas. I think it's worth trying some sort of tablegen op interface approach.

Reply:
!!

Looks like fold /is/ allowed to change the operation in-place, and in that case should return the (new?) operation?

From https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-the-fold-method :

> ///  2. They can mutate the operation in place, without changing anything else
> ///     in the IR. In this case, return the operation itself.

TIL, would this help? Have foldResult return the operation instead of the result Value?

Reply:
I don't think you can actually return the operation, `OpFoldResult` is declared as:
```cpp
class OpFoldResult : public PointerUnion<Attribute, Value>
```

Anyway, so far I've gotten tablegen to optionally define a `fold` method if `hasFolder` is `true`, that calls a `constFold` function meant to be implemented by the user then does all of the result checking. This currently works for single result ops but it would be cool if I could get it to work for single and mult-result ops, then we could put it on `FIRRTLOp` and not have to worry about needing to opt into it.

Reply:
Alright, I've updated `FIRRTLOp` to automatically handle folding type changes.